[lldb] Expand background symbol lookup

JDevlieghere opened the pull request:
LLDB has a setting (symbols.enable-background-lookup) that calls
dsymForUUID on a background thread for images as they appear in the
current backtrace. Originally, the laziness of only looking up symbols
for images in the backtrace only existed to bring the number of
dsymForUUID calls down to a manageable number.

Users have requesting the same functionality but blocking. This gives
them the same user experience as enabling dsymForUUID globally, but
without the massive upfront cost of having to download all the images,
the majority of which they'll likely not need.

This patch renames the setting to have a more generic name
(symbols.download) and changes its values from a boolean to an enum.
Users can now specify "off", "background" and "foreground". The default
remains "off" although I'll probably change that in the near future.

bulbazord commented on lldb/source/Core/CoreProperties.td:
```diff
@@ -5,10 +5,11 @@ let Definition = "modulelist" in {
     DefaultTrue,
     Desc<"Control the use of external tools and repositories to locate symbol files. Directories listed in target.debug-file-search-paths and directory of the executable are always checked first for separate debug info files. Then depending on this setting: On macOS, Spotlight would be also used to locate a matching .dSYM bundle based on the UUID of the executable. On NetBSD, directory /usr/libda ...
-  def EnableBackgroundLookup: Property<"enable-background-lookup", "Boolean">,
+  def LazySymbolLookup: Property<"lazy-lookup", "Enum">,
```
I think the name should probably be different. "LazySymbolLookup" being set to off makes me think it will look things up eagerly. Looking up the possible values of "off", "background", and "foreground" also don't really illustrate what they mean either.

I would propose the setting name "external-symbol-load-behavior" with the enumeration values "off", "background", and "blocking" or something to this effect.

JDevlieghere replied:
This is entirely orthogonal to the usual external symbol lookup. The latter uses DebugSymbols/Spotlight which may or may not use dsymForUUID under the hood. I considered on-demand but that's already taken by SymbolFileOnDemand. I'm open to other names but I don't have a better idea :-)

jasonmolenda replied:
I keep coming back to synonyms for "on demand" but not clear why they are different, like "as-needed-symbol-lookup".  Or "backtrace-driven-symbol-lookup".  

I don't know if there's any really clear name that could help a first-time user to distinguish between the different mechanisms.  We have the existing `symbols.load-on-demand` which is really avoiding loading the DWARF for any binaries until criteria are met (if I'm reading the docs/use/ondemand.rst correctly).  In this case, lldb *has* the DWARF for the binaries locally, it's just not reading it until the debug info is needed (with a few caveats like "breakpoints by name on inlined functions won't work").

This mechanism is instead "Download the DWARF (dSYM) and executable (if possible) from an external server, and load it immediately in to lldb".  For instance, you connect to a process on a remote system and none of the binaries running on the remote system, e.g. libsystem_c.dylib, exist on the host computer.  lldb can use this mechanism to download the correct libsystem_c.dylib and libsystem_c.dylib.dSYM to the host computer and load them in to lldb.  Instead of doing that for *all* binaries on connect, this allows us to avoid hitting the external servers until specific binaries are needed.

I might try to work 'symbol-download' into the setting name, that is a real difference between the two.

JDevlieghere replied:
I like your suggestion. @bulbazord How do you feel about `symbol-download={off, background,foreground}`?

jasonmolenda replied:
`symbols.download-on-demand` lol it's actually accurate but probably too overly clever compared to `symbols.load-on-demand`.

jasonmolenda replied:
(just to be clear, my fear of `symbols.download-on-demand` in addition to the existing `symbols.load-on-demand` is that no one except us will ever understand what the difference is.  I can sit here and say "it's very clear if you think about it!" but I think it's not worth it to be this subtle, IMO.)

jasonmolenda commented:
I like this change, this will be a very handy capability for people debugging corefiles & live sessions on remote computers when they have a mechanism to find the binaries and/or debug info for those processes, without needing to download all of them. One change I would also roll in in  `SymbolLocator::DownloadSymbolFileAsync()` is to request `copy_executable=true` here,

```
    if (!PluginManager::DownloadObjectAndSymbolFile(module_spec, error,
                                                    /*force_lookup=*/true,
                                                    /*copy_executable=*/false))
```

So we download both the dSYM and binary for situations where we don't have a local binary.

JDevlieghere committed:
[lldb] Expand background symbol lookup

LLDB has a setting (symbols.enable-background-lookup) that calls
dsymForUUID on a background thread for images as they appear in the
current backtrace. Originally, the laziness of only looking up symbols
for images in the backtrace only existed to bring the number of
dsymForUUID calls down to a manageable number.

Users have requesting the same functionality but blocking. This gives
them the same user experience as enabling dsymForUUID globally, but
without the massive upfront cost of having to download all the images,
the majority of which they'll likely not need.

This patch renames the setting to have a more generic name
(symbols.download) and changes its values from a boolean to an enum.
Users can now specify "off", "background" and "foreground". The default
remains "off" although I'll probably change that in the near future.

bulbazord reviewed (approved):
I'm happy w/ the name. Thanks!

clayborg commented on lldb/source/Core/CoreProperties.td:
```diff
@@ -5,10 +5,11 @@ let Definition = "modulelist" in {
     Global,
     DefaultTrue,
     Desc<"Control the use of external tools and repositories to locate symbol files. Directories listed in target.debug-file-search-paths and directory of the executable are always checked first for separate debug info files. Then depending on this setting: On macOS, Spotlight would be also used to locate a matching .dSYM bundle based on the UUID of the executable. On NetBSD, directory /usr/libda ...
-  def EnableBackgroundLookup: Property<"enable-background-lookup", "Boolean">,
```
Should we leave this setting around and adapt it to use the new setting with the correct defaults set? If someone has a `settings set enable-background-lookup true` in their ~/.lldbinit file, this will cause the .lldbinit file to stop executing lines if this setting fails to be set. This is kind of like a backward compatibility in our API kind of thing. We can make this setting as deprecated and tell people to use the `download` setting instead, but it would be nice if it still worked.

If this was used for dsymForUUID only and there weren't many customers it could be ok to remove, but I would err on the side of caution.

JDevlieghere replied:
I was on the fence, as I've been encouraging users to use  `settings set symbols.enable-background-lookup true` in their `.lldbinit` (because older versions don't know about the setting either). On the other hand it's very little work to keep the old setting working, so why not. I've pushed a commit that makes the old setting an alias.

clayborg commented on lldb/source/Core/CoreProperties.td:
```diff
@@ -5,10 +5,11 @@ let Definition = "modulelist" in {
-    Desc<"On macOS, enable calling dsymForUUID (or an equivalent script/binary) in the background to locate symbol files that weren't found.">;
+    DefaultEnumValue<"eSymbolDownloadOff">,
+    EnumValues<"OptionEnumValues(g_download_enum_values)">,
+    Desc<"On macOS, lazily download symbols with dsymForUUID (or an equivalent script/binary) as images appear in the current backtrace.">;
```
Is it truly only for backtraces? Seems like an address lookup in a shared library might be a good place to also get the symbols.

JDevlieghere replied:
For now yes, but we could extend this in the future.

clayborg commented on lldb/include/lldb/Core/ModuleList.h:
```diff
@@ -47,6 +47,26 @@ class UUID;
+    {
+        lldb::eSymbolDownloadBackground,
+        "background",
+        "Download symbols in the background for images as they appear in the "
+        "backtrace.",
```
Is this truly only for backtraces? Wouldn't an address lookup in a binary also qualify as a way to force debug info to be loaded?

medismailben reviewed (approved):
LGTM!

clayborg commented:
Does this setting only apply to the downloading of symbols that are auto downloaded for stack traces? Or does it apply to any other areas that auto download stuff? Do we want a setting to control the auto downloading of symbols to clarify? Something like:
```
(lldb) settings set symbols.auto-download-stack-symbols true
```
Right now your setting seems to be targeting stack auto symbol downloads, but this setting is pretty generic and could apply to other things in the future like auto downloading symbols when an address lookup happens on a module, finding the real definition for a type when we know a type lives in a shared library (like ObjC and C++ vtables can easily tell us).

JDevlieghere commented:
@clayborg The idea behind the feature is to get symbols for things that are relevant to the user. Right now, that's only hooked up for images that appear in the stack trace, but there are certainly other places this would be useful. So yeah, I absolutely expect this to apply to other things in the future as well. 

We currently check the setting right before downloading, so in a generic place that would be shared by all the other places that could trigger a download. Therefore I think the "mode" should be generic. If we have different triggers we could have separate settings to turn those things off (e.g. only download symbols for images in the stack trace, not for address lookups). So down the line it could look something like this:

```
(lldb) settings set symbols.auto-download background
(lldb) settings set symbols.auto-download-stack-symbols true
(lldb) settings set symbols.auto-download-address-lookup false
```

Let me know if you think it's critical to add `symbols.auto-download-stack-symbols` now in addition to `symbols.auto-download` (which I like better than `symbols.download`, thanks for the suggestion!) .

JDevlieghere commented:
Rebased & renamed the property to `symbols.auto-download`. I think all the review comments have been addressed. The only oustanding question is if @clayborg wants to have an additional setting for `symbols.auto-download-stack-symbols`.

clayborg commented:
> @clayborg The idea behind the feature is to get symbols for things that are relevant to the user. Right now, that's only hooked up for images that appear in the stack trace, but there are certainly other places this would be useful. So yeah, I absolutely expect this to apply to other things in the future as well.
> 
> We currently check the setting right before downloading, so in a generic place that would be shared by all the other places that could trigger a download. Therefore I think the "mode" should be generic. If we have different triggers we could have separate settings to turn those things off (e.g. only download symbols for images in the stack trace, not for address lookups). So down the line it could look something like this:
> 
> ```
> (lldb) settings set symbols.auto-download background
> (lldb) settings set symbols.auto-download-stack-symbols true
> (lldb) settings set symbols.auto-download-address-lookup false
> ```
> 
> Let me know if you think it's critical to add `symbols.auto-download-stack-symbols` now in addition to `symbols.auto-download` (which I like better than `symbols.download`, thanks for the suggestion!) .

Thanks for the update. Not critical to this patch, just wondering where we are going with this in the future and auto downloading of symbols.

nico commented:
This doesn't build on Linux: http://45.33.8.238/linux/130329/step_4.txt

Please take a look and revert for now if it takes a while to fix.

(Looks like you added a definition on non apple; maybe the declaration is just still missing)
